src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs

```diff
@@ -65,10 +67,45 @@ private void GetInitialDiagnostics(CancellationToken cancellationToken)
                     var workspace = project.Solution.Workspace;
                     foreach (var updateArgs in _owner._diagnosticService.GetDiagnosticsUpdatedEventArgs(workspace, project.Id, document.Id, cancellationToken))
                     {
-                        var diagnostics = _owner._diagnosticService.GetDiagnostics(updateArgs.Workspace, updateArgs.ProjectId, updateArgs.DocumentId, updateArgs.Id, includeSuppressedDiagnostics: false, cancellationToken: cancellationToken);
-                        OnDiagnosticsUpdated(DiagnosticsUpdatedArgs.DiagnosticsCreated(updateArgs.Id, updateArgs.Workspace, project.Solution, updateArgs.ProjectId, updateArgs.DocumentId, diagnostics.AsImmutableOrEmpty()));
+                        var diagnostics = GetInitialDiagnostics(project.Solution, updateArgs, cancellationToken);
+                        if (diagnostics.Length == 0)
+                        {
+                            continue;
+                        }
+
+                        OnDiagnosticsUpdated(
+                            DiagnosticsUpdatedArgs.DiagnosticsCreated(
+                                updateArgs.Id, updateArgs.Workspace, project.Solution, updateArgs.ProjectId, updateArgs.DocumentId, diagnostics));
+                    }
+                }
+            }
+
+            private ImmutableArray<DiagnosticData> GetInitialDiagnostics(Solution solution, UpdatedEventArgs args, CancellationToken cancellationToken)
```

Comment:
I would call this "AdjustInitialDiagnostics".  And explain in the code that hte initial diagnostics may not match the source text.